addons/product/tests/test_variants.py

```diff
@@ -461,3 +461,46 @@ def test_07_create_no_create_dynamic_always(self):
         self.assertEqual(len(toto.attribute_line_ids.mapped('attribute_id')), 10)
         self.assertEqual(len(toto.attribute_line_ids.mapped('value_ids')), 100)
         self.assertEqual(len(toto.product_variant_ids), 0)
+
+class TestVariantsImages(common.TestProductCommon):
+
+    def setUp(self):
+        res = super(TestVariantsImages, self).setUp()
+
+        # To make the test independent of the demo data, you only need to rewrite this line
+        self.template = self.env.ref('product.product_product_4_product_template')
+        self.img = self.template.image
+        self.variants = self.template.product_variant_ids.sorted(lambda p: p.id)
+        self.images = self.variants.mapped('image')
+        self.img2 = self.images[2]
+
+        return res
+
+    def test_variant_images(self):
+        """Check that on variant, the image used is the image_variant if set,
+        and defaults to the template image otherwise.
+        """
+        # all images are different
+        self.assertEqual(len(set(self.images)), 4)
+        # the first has no image variant, all the others do
+        self.assertFalse(self.variants[0].image_variant)
+        self.assertFalse(False in self.variants[1:].mapped('image_variant'))
+
+        # template image is the same as this one, since it has no variant
+        self.assertEqual(self.img, self.images[0])
```

Comment:
Maybe `self.assertEqual(self.variant[0].image, self.template.image)` is more clear what is tested (or rename the variables).